Add codec and processor combinations to S3 source

Naarcha-AWS opened the pull request:
### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+# Codec processor combinations
+
+At ingestion time, data received by the [`s3 source`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by codecs, such as [Apache Avro](https://avro.apache.org/docs/1.11.1/specification/). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site.baseurl}}/d ...
```
We should only format "s3" as code, not "s3 source." Basically, let's move the back quotes. Thus, this should be:

> by the [`s3` source]

dlvenable replied:
> such as [Apache Avro]

I recommend starting with a codec like JSON over Avro. JSON is probably the most popular codec.

Naarcha-AWS replied:
```suggestion
At ingestion time, data received by the [s3 source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by codecs, such as [Apache Avro](https://avro.apache.org/docs/1.11.1/specification/). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+| **Compression type** | **Codec** | **Processor** |
+|---|---|---|
+| None /gzip/ Automatic |[JSON]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#json-codec) | [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) | 
```
Do we need the compression type column? That is unrelated to the processor/codec.

dlvenable replied:
> [JSON]

Please make this lowercase:

> [json]

dlvenable replied:
When using the `json` codec, the user does not need to use the `parse_json` processor. Indeed, for most use-cases, the user shouldn't use `parse_json`.

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+| **Compression type** | **Codec** | **Processor** |
+|---|---|---|---|
+| None /gzip/ Automatic | [Newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec)  | 
```
> [Newline]

This should be all lowercase:

> [newline]

dlvenable replied:
When using the `newline` to read ND-JSON data, users need to use the `parse_json` processor. So that should be added here.

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+| **Compression type** | **Codec** | **Processor** | **Limitations**
+|---|---|---|---|---|
+| None /gzip/ Automatic | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) | None | Header is automatically detected and used for index mapping with the CSV codec. |
```
> [CSV]

This should be lowercase:

> [csv]

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+| **Compression type** | **Codec** | **Processor** | **Limitations**
+|---|---|---|---|---|
+| None /gzip/ Automatic | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) | None | Header is automatically detected and used for index mapping with the CSV codec. |
+| None /gzip/ Automatic | [Newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) | [CSV]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) | Header is detected if `header_destination` is configured in the `newline` codec. <br>The header can be used in the CSV processor under `column_names_source_key`. |
```
Again, let's be consistent with the case:

> [Newline]

to:

> [newline]

And:

> [CSV]

to:

> [csv]

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+The CSV input type is most effective with the following codec processor combinations.
+
+| **Compression type** | **Codec** | **Processor** | **Limitations**
```
I wonder if the table format is the right way to convey this to users. It seems that we could have the following instead.

1. Some example data
2. A paragraph on the data and the configuration
3. A sample pipeline configuration, or perhaps just a partial configuration.

vagimeli replied:
@Naarcha-AWS Did this comment from @dlvenable get addressed? He needs to write this section based on his comments, as they are technical in nature and he is the SME.

dlvenable commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+| **Compression type** | **Processor** |
+|---|---|
+| None /gzip/ Automatic | [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
```
I think that a table may not convey the best way to convey this. For one, using the `parquet` codec versus using `s3_select` is a decision that may require some additional consideration. And the way you configure `s3_select` is different.

Naarcha-AWS replied:
What additional considerations?

dlvenable commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -149,6 +149,22 @@ Option | Required | Type   | Description
+
+At ingestion time, data received by the S3 source can be parsed by codecs. However, codecs only support a particular type of data input type, compressions types, and processors. The following table describes each codec, data type, and processor combination where performance of the pipeline can be improved. 
+
+| **Input Data Type** | **Compression type** | **Codec** | **Processor** | Notes |
```
Rather than adding a table here, perhaps we can link to the other page for users.

Naarcha-AWS replied:
I'm going to remove the table.

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,59 @@
+
+| **Compression type** | **Codec** | **Processor** |
+|---|---|---|
+| None /gzip/ Automatic |[JSON]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#json-codec) | [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) | 
```
Do we have JSON processor besides `parse_JSON`?

dlvenable replied:
No, `parse_json` is the only JSON processor we have.

Naarcha-AWS replied:
So do they need a processor at all when using the JSON codec?

dlvenable replied:
When using the `json` codec, users don't need a processor. Their data will already be somewhat ready.

Keep in mind that customers may wish to use any number of processors after the data has been parsed to normalize data, remove data, etc.

Naarcha-AWS commented:
@dlvenable: Addressed most of your feedback. I'll need some help developing examples, but if we want to marge this now before we add them we can circle back around to the examples.

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+- [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
+- [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source) |
+
+## Avro
```
```suggestion
## Avro
```

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+## Avro
+
+[Apace Avro] helps streamline streaming data pipelines. It is most efficient when used with the [Avro]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) inside an `s3` sink.
```
```suggestion
[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) codec inside an `s3` sink.
```

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+The [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec parses each row as a single log event. However, the codec will only detect a header when `header_destination` is configured. Then, the [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) processor outputs the event into columns. The header detected i ...
+
+
+## Parquet
```
```suggestion
## parquet
```

vagimeli replied:
Leave the headings capitalized. Use code font in the body text.

Naarcha-AWS replied:
Parquet and Avro are the exceptions to the code rule since they represent a data storage format, not something that appears in code.

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
```
```suggestion
[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
```

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when paired with the following processors depending on the compression type.
+
+
+- [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
```
```suggestion

```

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+
+- [Parquet]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#parquet-codec) |
+- [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source) |
```
```suggestion

```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+# Codec processor combinations
+
+At ingestion time, data received by the [`s3` source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by [codecs]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#codec). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site. ...
```
```suggestion
At ingestion time, data received by the [`s3` source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by [codecs]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#codec). Codecs compresses and decompresses large data sets in a certain format before ingestion them through a Data Prepper pipeline [processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/processors/).
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+At ingestion time, data received by the [`s3` source]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3/) can be parsed by [codecs]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#codec). which will compress and decompress large sets of data in a certain format before ingestion through a Data Prepper pipelines [processor]({{site.url}}{{site. ...
+
+While most codecs can be used with most processors, the following codec processor combinations can help make your pipeline more efficient when using the following input types.
```
```suggestion
While most codecs can be used with most processors, the following codec processor combinations can make your pipeline more efficient when used with the following input types.
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+While most codecs can be used with most processors, the following codec processor combinations can help make your pipeline more efficient when using the following input types.
+
+## JSON Array
```
```suggestion
## JSON array
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,55 @@
+
+## NDJSON
+
+Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson), each row of data can be delimited by a newline, which means data is processed per-line as opposed to an array.
```
```suggestion
Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson) allows for each row of data to be delimited by a newline, meaning data is processed per line instead of an array.
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson), each row of data can be delimited by a newline, which means data is processed per-line as opposed to an array.
+
+The NDJSON input type requires is parsed using the [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec, which parses each single line as a single log event. The [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) processor then outputs each line as a single event.
```
```suggestion
The NDJSON input type is parsed using the [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec, which parses each single line as a single log event. The [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) processor then outputs each line as a single event.
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+Unlike a JSON array, [NDJSON](https://www.npmjs.com/package/ndjson), each row of data can be delimited by a newline, which means data is processed per-line as opposed to an array.
+
+The NDJSON input type requires is parsed using the [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec, which parses each single line as a single log event. The [parse_json]({{site.url}}{{site.baseurl}}data-prepper/pipelines/configuration/processors/parse-json/) processor then outputs each line as a single event.
+
```
Delete extra blank lines

vagimeli replied:
@Naarcha-AWS The comments don't appear to be addressed. This file has several sections with extra blank lines that need to be deleted so that this document is in proper format.

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+## CSV
+
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processors or the `csv` codec.
```
```suggestion
The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processor or the `csv` codec.
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+The CSV data type inputs data as a table. It can used without a codec or processor, but it does require one or the other, for example, either just the `csv` processors or the `csv` codec.
+
+The CSV input type is most effective with the following codec/processor combinations.
```
```suggestion
The CSV input type is most effective when used with the following codec processor combinations.
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+The CSV input type is most effective with the following codec/processor combinations.
+
+
+### csv codec
```
```suggestion
### CSV codec
```

Naarcha-AWS replied:
This is meant to match the codec appears in the code. Its the standard the Data Prepper team wishes to use and is used across the entire Data Prepper guide.

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+### csv codec
+
+The [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) codec without a processor. When used without a processor, the `csv` codec automatically detects headers from the CSV and uses them for index mapping.
```
```suggestion
When the [`csv` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) is used without a processor, it automatically detects headers from the CSV and uses them for index mapping.
```

vagimeli replied:
Should this read "...the CSV **file** and..."

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+The [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) codec without a processor. When used without a processor, the `csv` codec automatically detects headers from the CSV and uses them for index mapping.
+
+### newline codec 
```
```suggestion
### Newline codec 
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+### newline codec 
+
+The [newline]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) codec parses each row as a single log event. However, the codec will only detect a header when `header_destination` is configured. Then, the [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) processor outputs the event into columns. The header detected i ...
```
```suggestion
The [`newline` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#newline-codec) parses each row as a single log event. The codec will only detect a header when `header_destination` is configured. The processor then [csv]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/csv/) outputs the event into columns. The header detected in `header_destination` from the `newline` codec can be used in the `csv` processor under `column_names_source_key.`
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec, however positive results can be achieved when configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
```
```suggestion
[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec. Positive results, however, can be achieved when it's configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+## Avro
+
+[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro`]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) codec inside an `s3` sink.
```
```suggestion
[Apache Avro] helps streamline streaming data pipelines. It is most efficient when used with the [`avro` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sinks/s3#avro-codec) inside an `s3` sink.
```

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
 
 ## codec
```
```suggestion
## Codec
```

Naarcha-AWS replied:
See previous comment about headers.

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
 ## codec
 
-The `codec` determines how the `s3` source parses each S3 object.
+The `codec` determines how the `s3` source parses each S3 object. For increased and more efficient performance, you can use codecs in [combination]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
```
```suggestion
The `codec` determines how the `s3` source parses each Amazon S3 object. For increased and more efficient performance, you can use [codec combinations]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
```

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -121,7 +121,7 @@ Option | Required | Type | Description
-The `codec` determines how the `s3` source parses each S3 object.
+The `codec` determines how the `s3` source parses each S3 object. For increased and more efficient performance, you can use codecs in [combination]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/codec-processor-combinations/) with certain processors.
 
 ### newline codec
```
```suggestion
### Newline codec
```

Naarcha-AWS replied:
See previous comments about headers in the Data Prepper section.

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -149,6 +149,9 @@ Option | Required | Type   | Description
 `header` | No   | String list | The header containing the column names used to parse CSV data.
 `detect_header` | No   | Boolean   | Whether the first line of the S3 object should be interpreted as a header. Default is `true`.
```
```suggestion
`detect_header` | No   | Boolean   | Whether the first line of the Amazon S3 object should be interpreted as a header. Default is `true`.
```

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -149,6 +149,9 @@ Option | Required | Type   | Description
+
 ## Using `s3_select` with the `s3` source<a name="s3_select"></a>
 
 When configuring `s3_select` to parse S3 objects, use the following options.
```
```suggestion
When configuring `s3_select` to parse Amazon S3 objects, use the following options:
```

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
 `interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). 
```
```suggestion
`interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO 8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). 
```

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
 `interval` | Yes  | String | Indicates the minimum interval between each scan. The next scan in the interval will start after the interval duration from the last scan ends and when all the objects from the previous scan are processed. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). 
 `count` | No | Integer | Specifies how many times a bucket will be scanned. Defaults to `Integer.MAX_VALUE`.
```
```suggestion
`count` | No | Integer | Specifies how many times a bucket is scanned. Defaults to `Integer.MAX_VALUE`.
```

Naarcha-AWS replied:
The `count` doesn't occur until the pipeline runs, hence the tense.

vagimeli commented on _data-prepper/pipelines/configuration/sources/s3.md:
```diff
@@ -225,6 +228,7 @@ Option | Required | Type | Description
+
 ## Metrics
 
 The `s3` source includes the following metrics.
```
```suggestion
The `s3` source includes the following metrics:
```

Naarcha-AWS replied:
I was told by @natebower to make this a period.

vagimeli reviewed (commented):
@Naarcha-AWS Editorial review is complete. Please see edits and ping me directly once they've been addressed so that I can approve for merging. Thanks!

dlvenable reviewed (approved):
I think we should include some examples in the `codec-processor-combinations` file, but we can follow on with this.

Naarcha-AWS commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+
+When the [`csv` codec]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#csv-codec) is used without a processor, it automatically detects headers from the CSV and uses them for index mapping.
+
+### Newline codec 
```
```suggestion
### `newline` codec 
```

vagimeli commented on _data-prepper/common-use-cases/codec-processor-combinations.md:
```diff
@@ -0,0 +1,53 @@
+## Parquet
+
+[Apache Parquet](https://parquet.apache.org/docs/overview/) is a columnar storage format built for Hadoop. It is most efficient without the use of a codec. Positive results, however, can be achieved when it's configured with [S3 Select]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/sources/s3#using-s3_select-with-the-s3-source).
+
```
@Naarcha-AWS Delete extra blank lines

vagimeli reviewed (approved):
@Naarcha-AWS Check the file to delete the extra blank lines so that it's properly formatted before merging. Thanks!

vagimeli commented:
@Naarcha-AWS If you all add more text in the examples, those changes need to go through editorial review.
